### PR TITLE
Fix view history

### PR DESCRIPTION
**Pull request by Stezido**
This fixes should solve #87 and #85. I am aiming for a complete and readable history where items are shown in the right direction from new to old.

### CODE REVIEW

**Comment by mathiashoeld** [comment]
@Stezido this does not seem to work properly yet. after provisioning, the history cannot be loaded.

File: api/src/project_view_details.ts [hunk 1]
```diff
@@ -77,11 +77,13 @@ function mkSwaggerSchema(server: FastifyInstance) {
                       type: "array",
                       items: {
                         type: "object",
+                        additionalProperties: true,
```
Comment (kevinbader) [inline api/src/project_view_details.ts]: why?

File: api/src/project_view_history.ts [hunk 2]
```diff
@@ -11,6 +11,7 @@ import { BusinessEvent } from "./service/domain/business_event";
 import { ServiceUser } from "./service/domain/organization/service_user";
 import * as Project from "./service/domain/workflow/project";
 import * as Subproject from "./service/domain/workflow/subproject";
+import logger from "./lib/logger";
```
Comment (kevinbader) [inline api/src/project_view_history.ts]: not sorted

Comment (kevinbader) [reply api/src/project_view_history.ts]: and below (can't comment there) the request body related stuff is dead code - the request can't have a body

File: api/src/project_view_history.ts [hunk 3]
```diff
@@ -201,7 +204,6 @@ export function addHttpHandler(server: FastifyInstance, urlPrefix: string, servi
         for (const subproject of subprojects) {
           events.push(...subproject.log);
         }
-
```
Comment (kevinbader) [inline api/src/project_view_history.ts]: why? the lines above add the subproject log to the project log, while the line below sorts _all_ log events.. I'd keep the empty line to separate these two things

File: api/src/service/cache2.ts [hunk 4]
```diff
@@ -443,20 +444,25 @@ function addEventsToCache(cache: Cache2, streamName: string, newEvents: Business
...
+      // Do nothing, because informations will be reflected in aggregates
       break;
   }
 }
 
 export function updateAggregates(ctx: Ctx, cache: Cache2, newEvents: BusinessEvent[]) {
-  // we ignore the errors
-  const { projects } = sourceProjects(ctx, newEvents, cache.cachedProjects);
+  const { projects, errors: pErrors = [] } = sourceProjects(ctx, newEvents, cache.cachedProjects);
+  if (!isEmpty(pErrors)) logger.warn("sourceProject caused error: ", pErrors);
```
Comment (kevinbader) [inline api/src/service/cache2.ts]: if there is an ill-formatted event in the history (for whatever reason), there will be a _warn_ printed _each time the uggregates are built up_. This is most likely not what we want here. Instead, we should either log `debug` level instead, or ignore the errors here altogether (like we did so far).

File: api/src/service/cache2.ts [hunk 5]
```diff
@@ -443,20 +444,25 @@ function addEventsToCache(cache: Cache2, streamName: string, newEvents: Business
...
     cache.cachedProjects.set(project.id, project);
   }
 
-  const { subprojects } = sourceSubprojects(ctx, newEvents, cache.cachedSubprojects);
+  const { subprojects, errors: spErrors = [] } = sourceSubprojects(
+    ctx,
+    newEvents,
+    cache.cachedSubprojects,
+  );
+  if (!isEmpty(spErrors)) logger.warn("sourceSubproject caused error: ", spErrors);
```
Comment (kevinbader) [inline api/src/service/cache2.ts]: see comment above

File: api/src/service/cache2.ts [hunk 6]
```diff
@@ -467,7 +473,12 @@ export function updateAggregates(ctx: Ctx, cache: Cache2, newEvents: BusinessEve
       : lookUp.add(subproject.id);
   }
 
-  const { workflowitems, errors } = sourceWorkflowitems(ctx, newEvents, cache.cachedWorkflowItems);
+  const { workflowitems, errors: wErrors = [] } = sourceWorkflowitems(
+    ctx,
+    newEvents,
+    cache.cachedWorkflowItems,
+  );
+  if (!isEmpty(wErrors)) logger.warn("sourceWorkflowitems caused error: ", wErrors);
```
Comment (kevinbader) [inline api/src/service/cache2.ts]: see comment above

File: api/src/service/domain/workflow/subproject_eventsourcing.ts [hunk 7]
```diff
@@ -31,9 +32,9 @@ export function sourceSubprojects(
     if (!event.type.startsWith("subproject_")) {
       continue;
     }
-
     const result = applySubprojectEvent(ctx, subprojects, event);
     if (Result.isErr(result)) {
+      logger.warn(result);
```
Comment (kevinbader) [inline api/src/service/domain/workflow/subproject_eventsourcing.ts]: please remove - we don't want eventsourcing to produce any logs, that's the reason it returns the errors instead

File: api/src/service/subproject_permission_grant.ts [hunk 8]
```diff
@@ -39,7 +40,6 @@ export async function grantSubprojectPermission(
     ),
   );
   if (Result.isErr(result)) return Promise.reject(result);
-
```
Comment (kevinbader) [inline api/src/service/subproject_permission_grant.ts]: above: get result, check result, return if result is error

below: store the events that got generated by the domain layer

why not keep them separated?

File: api/src/subproject_view_details.ts [hunk 9]
```diff
@@ -91,63 +91,31 @@ function mkSwaggerSchema(server: FastifyInstance) {
                       type: "array",
                       items: {
                         type: "object",
+                        additionalProperties: true,
```
Comment (kevinbader) [inline api/src/subproject_view_details.ts]: why?

File: frontend/src/pages/SubProjects/reducer.js [hunk 10]
```diff
@@ -137,10 +137,11 @@ export default function detailviewReducer(state = defaultState, action) {
       return state.set("isHistoryLoading", true);
     case FETCH_PROJECT_HISTORY_SUCCESS:
       return state.merge({
-        historyItems: [...state.get("historyItems"), ...fromJS(action.events)],
+        historyItems: [...fromJS(action.events), ...state.get("historyItems")],
```
Comment (kevinbader) [inline frontend/src/pages/SubProjects/reducer.js]: after that, `historyItems` isn't immutable, which is probably not what was intended.

```suggestion
        historyItems: fromJS(action.events).concat(state.get("historyItems")),
```

File: frontend/src/pages/Workflows/reducer.js [hunk 11]
```diff
@@ -287,15 +287,17 @@ export default function detailviewReducer(state = defaultState, action) {
       return state.set("isHistoryLoading", true);
     case FETCH_SUBPROJECT_HISTORY_SUCCESS:
       return state.merge({
-        historyItems: [...state.get("historyItems"), ...fromJS(action.events)],
+        historyItems: [...fromJS(action.events), ...state.get("historyItems")],
```
Comment (kevinbader) [inline frontend/src/pages/Workflows/reducer.js]: see comment above (tl;dr: use `.concat`)

File: api/src/project_view_details.ts (hunk 1 above)
Comment (Stezido) [reply api/src/project_view_details.ts]: you are right, TraceEvents can only look like this and do not allow any other properties

File: api/src/service/domain/workflow/subproject_get.ts [hunk 12]
```diff
@@ -46,6 +45,7 @@ const requiredPermissions = new Map<EventType, Intent[]>([
   ["subproject_archived", ["subproject.viewSummary", "subproject.viewDetails"]],
   ["subproject_projected_budget_updated", ["subproject.viewDetails"]],
   ["subproject_projected_budget_deleted", ["subproject.viewDetails"]],
+  ["subproject_items_reordered", ["subproject.reorderWorkflowitems"]],
```
Comment (kevinbader) [inline api/src/service/domain/workflow/subproject_get.ts]:
```suggestion
  ["workflowitems_reordered", ["subproject.reorderWorkflowitems"]],
```

File: api/src/service/domain/workflow/subproject_list.ts [hunk 13]
```diff
@@ -49,6 +49,7 @@ const requiredPermissions = new Map<EventType, Intent[]>([
   ["subproject_archived", ["subproject.viewSummary", "subproject.viewDetails"]],
   ["subproject_projected_budget_updated", ["subproject.viewDetails"]],
   ["subproject_projected_budget_deleted", ["subproject.viewDetails"]],
+  ["subproject_items_reordered", ["subproject.reorderWorkflowitems"]],
```
Comment (kevinbader) [inline api/src/service/domain/workflow/subproject_list.ts]:
```suggestion
  ["workflowitems_reordered", ["subproject.reorderWorkflowitems"]],
```

File: api/src/service/domain/workflow/workflowitem_list.ts [hunk 14]
```diff
@@ -65,7 +65,7 @@ const requiredPermissions = new Map<EventType, Intent[]>([
   ["workflowitem_assigned", ["workflowitem.view"]],
   ["workflowitem_updated", ["workflowitem.view"]],
   ["workflowitem_closed", ["workflowitem.view"]],
-  ["workflowitem_reordered", ["workflowitem.view"]],
+  ["subproject_items_reordered", ["workflowitem.view"]],
```
Comment (kevinbader) [inline api/src/service/domain/workflow/workflowitem_list.ts]:
```suggestion
  ["workflowitems_reordered", ["workflowitem.view"]],
```

File: frontend/src/pages/SubProjects/ProjectHistoryContainer.js [hunk 15]
```diff
@@ -21,46 +21,56 @@ const calculateHistory = items => {
...
+      );
+    case "subproject_permission_revoked":
       return formatString(
         strings.history.subproject_revokePermission_details,
         createdBy,
         formatPermission(data),
         data.identity,
         snapshot.displayName
       );
+    case "subproject_items_reordered":
```
Comment (kevinbader) [inline frontend/src/pages/SubProjects/ProjectHistoryContainer.js]:
```suggestion
    case "workflowitems_reordered":
```

File: frontend/src/pages/Workflows/SubProjectHistoryContainer.js [hunk 16]
```diff
@@ -22,56 +22,58 @@ const calculateHistory = items => {
...
       return formatString(strings.history.subproject_assign, createdBy, snapshot.displayName, data.identity);
-    case "workflowitem.close":
-      return formatString(strings.history.workflowitem_close, createdBy, snapshot.displayName);
-    case "subproject.close":
+    case "subproject_updated":
+      return formatString(strings.history.subproject_update, createdBy, snapshot.displayName);
+    case "subproject_closed":
       return formatString(strings.history.subproject_close, createdBy, snapshot.displayName);
-    case "subproject.intent.grantPermission":
+    case "subproject_items_reordered":
```
Comment (kevinbader) [inline frontend/src/pages/Workflows/SubProjectHistoryContainer.js]:
```suggestion
    case "workflowitems_reordered":
```

**Comment by Stezido** [comment]
Closes #85 
Closes #87

**Pull request merged**